cumulus/parachains/runtimes/assets/migrations/src/v1/mod.rs

```diff
@@ -0,0 +1,127 @@
+
+//! V1 migration.
+//!
+//! This migration is meant to upgrade the XCM version of asset locations from V3 to V4.
+//! It's only needed if the `AssetId` for this pallet is `VersionedLocation`
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarks;
+
+mod weights;
+
+use frame_support::{
+	migrations::{MigrationId, SteppedMigration, SteppedMigrationError},
+	pallet_prelude::PhantomData,
+	weights::WeightMeter,
+};
+use pallet_assets::{Asset, Config};
+use xcm::{v3, v4};
+
+pub const PALLET_MIGRATIONS_ID: &[u8; 13] = b"pallet-assets";
+
+/// Storage aliases for on-chain storage types before running the migration.
+pub mod old {
+	use super::{v3, Config};
+	use frame_support::{storage_alias, Blake2_128Concat};
+	use pallet_assets::{AssetDetails, DepositBalanceOf, Pallet};
+
+	pub type AssetDetailsOf<T, I> = AssetDetails<
+		<T as Config<I>>::Balance,
+		<T as frame_system::Config>::AccountId,
+		DepositBalanceOf<T, I>,
+	>;
+
+	/// The storage item we are migrating from.
+	#[storage_alias]
+	pub(super) type Asset<T: Config<I>, I: 'static> =
+		StorageMap<Pallet<T, I>, Blake2_128Concat, v3::Location, AssetDetailsOf<T, I>>;
+}
+
+pub struct Migration<T, I, W>(PhantomData<(T, I, W)>);
+impl<T: Config<I>, I: 'static, W: weights::WeightInfo> SteppedMigration for Migration<T, I, W>
+where
+	<T as Config<I>>::AssetId: From<v4::Location>,
+{
+	type Cursor = v3::Location;
+	type Identifier = MigrationId<13>;
+
+	fn id() -> Self::Identifier {
+		MigrationId { pallet_id: *PALLET_MIGRATIONS_ID, version_from: 0, version_to: 1 }
+	}
+
+	fn step(
+		mut cursor: Option<Self::Cursor>,
+		meter: &mut WeightMeter,
+	) -> Result<Option<Self::Cursor>, SteppedMigrationError> {
+		let required = W::conversion_step();
+		if meter.remaining().any_lt(required) {
+			return Err(SteppedMigrationError::InsufficientWeight { required });
+		}
+
+		loop {
+			if meter.try_consume(required).is_err() {
+				break;
+			}
+
+			cursor = Self::conversion_step(cursor);
+
+			if cursor.is_none() {
+				break;
+			}
+		}
+
+		Ok(cursor)
+	}
+}
+
+impl<T: Config<I>, I: 'static, W> Migration<T, I, W>
+where
+	<T as Config<I>>::AssetId: From<v4::Location>,
+{
+	pub fn conversion_step(cursor: Option<v3::Location>) -> Option<v3::Location> {
+		let mut iter = if let Some(last_key) = cursor {
+			// If a cursor is provided, start iterating from the value corresponding
+			// to the last key processed in the previous step of the migration.
+			old::Asset::<T, I>::iter_from(old::Asset::<T, I>::hashed_key_for(last_key))
+		} else {
+			// If there is no cursor, start iterating from the beginning.
+			old::Asset::<T, I>::iter()
+		};
+
+		if let Some((key, value)) = iter.next() {
+			// Most likely all locations will be able to be converted, but if they can't
+			// we log them to try again later.
+			let maybe_new_key: Result<v4::Location, _> = key.try_into();
+			if let Ok(new_key) = maybe_new_key {
+				old::Asset::<T, I>::remove(&key);
```

Comment:
I am not sure if it is safe to use an iterator `.next` *while* removing elements from the map at the same time or if it breaks the iterator...  
Could be safer to have a second step afterwards where it removes them, @cheme do you know?

Reply:
I am late at the party, but code was written to support this (calling next on a state that change: either adding or removing content).